process_report/process_report.py

```diff
@@ -1,9 +1,97 @@
 import argparse
 import os
+import sys
 
+import json
 import pandas
 
 
+### Invoice field names
+INVOICE_DATE_FIELD = 'Invoice Month'
+PROJECT_FIELD = 'Project - Allocation'
+PROJECT_ID_FIELD = 'Project - Allocation ID'
+PI_FIELD = 'Manager (PI)'
+INVOICE_EMAIL_FIELD = 'Invoice Email'
+INVOICE_ADDRESS_FIELD = 'Invoice Address'
+INSTITUTION_FIELD = 'Institution'
+INSTITUTION_ID_FIELD = 'Institution - Specific Code'
+SU_HOURS_FIELD = 'SU Hours (GBhr or SUhr)'
+SU_TYPE_FIELD = 'SU Type'
+COST_FIELD = 'Cost'
+CREDIT_FIELD = 'Credit'
+CREDIT_CODE_FIELD = 'Credit Code'
+BALANCE_FIELD = 'Balance'
+###
+
+
+def apply_credits_0001(dataframe):
+    credit_code = "0001"
+    pass
+
+
+def apply_credits_0002(dataframe):
+    """Applies the New PI Credit. This credit function expects the 
+    env var `C0002_OLD_PI` to be set, pointing to a txt file containing old PIs"""
```

Comment:
@knikolla When deciding how to pass arguments to the credit functions, I thought that there wasn't a simple way without using env vars. My thought process went like this:

- There is the assumption that we will have use and discontinue more credits in the future. These credits may require different parameters of varying amounts.  
- This assumption influenced how I designed the credit functions and `add_credits()`. The credit functions only take the invoice dataframe as the only function argument, which would allow `add_credits()` to simply iterate through the credit functions, instead of having to make an `if` block for each credit function, then pass in its specific parameters, like this:
```
def add_credits(dataframe, args):

    # Insert rows
    dataframe.insert(dataframe.columns.get_loc(COST_FIELD) + 1, CREDIT_FIELD, 0.0)
    dataframe.insert(dataframe.columns.get_loc(CREDIT_FIELD) + 1, CREDIT_CODE_FIELD, None)
    dataframe.insert(dataframe.columns.get_loc(CREDIT_CODE_FIELD) + 1, BALANCE_FIELD, 0.0)

    # Apply credits
    if credit_is_active(0001):
        dataframe = apply_credits_0001(dataframe, args.C0001)
    if credit_is_active(0002):
        dataframe = apply_credits_0002(dataframe, args.C0002_old_pi)
    if credit_is_active(0003):
        ...
```
- I thought that as more credits come and go, writing `add_credits` like this would just add another section of `process_report.py` that needs to be continually updated. Although not much work would be needed to update this function, it is still work nonetheless. 
- Given all this, I decided to pick the first approach. With all credit functions taking one arguement, the credit-specific parameters would have to come from env vars.

Reply:
I understand your reasoning, however:

- It doesn't have to be an argument to the function, there are ways to avoid that. Example: global variables, configuration options accessible globally. My ask here was for this to be an argument to the command line as this makes the application much simpler to use as opposed to an environment variable.
- In general, my preference is to try to avoid creating an abstraction or generalization when you have only one case to work with. We don't know what may extra information will be required to apply other kinds of credits and we will not have any other kinds of credits in the near term. It is extremely likely that whatever generalization you come up with here will not work for the second case of credits anyway. Start simple.

Reply:
I've restructured the code and added `old-pi-file` as a command arguement